src/core/models.py

```diff
@@ -281,6 +281,11 @@ def tls(self) -> bool:
         """Flag to check if TLS is enabled for the cluster."""
         return self.relation_data.get("tls", "") == "enabled"
 
+    @property
+    def s3_credentials(self) -> dict[str, str]:
```

Comment:
**suggestion** what about being a bit more specific than a `dict`? could we use an object? In Spark I believe we parse this into a `S3Connection` object, such that we can be a bit more explicit and type-safe with attributes

Reply:
Ok, I finally took the plunge and added a `stubs.py` module for custom types and definition overrides. Its very first member is `S3ConnectInfo`. Contrary to spark, this is not (yet) a pydantic model with runtime checking though

Added in c67a0e7e9f12c2092637ac41402405a34b6964a5